src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs

```diff
@@ -0,0 +1,308 @@
+        /// </remarks>
+        private T _item2;
+
+        /// <summary>
+        /// The fourth inline element.
+        /// </summary>
+        /// <remarks>
+        /// This field is accessed by offset from <see cref="_item0"/>. See <see cref="_item0"/> for usage information.
+        /// </remarks>
+        private T _item3;
+#pragma warning restore CS0649 // Field 'name' is never assigned to, and will always have its default value
+#pragma warning restore IDE0044 // Add readonly modifier
+
+        /// <summary>
+        /// The number of inline elements held in the array. This value is only used when <see cref="_builder"/> is
+        /// <see langword="null"/>.
+        /// </summary>
+        private int _count;
+
+        /// <summary>
+        /// A builder used for dynamic storage of collections that may exceed the limit for inline elements.
+        /// </summary>
+        /// <remarks>
+        /// This field is initialized to non-<see langword="null"/> the first time the <see cref="TemporaryArray{T}"/>
+        /// needs to store more than four elements. From that point, <see cref="_builder"/> is used instead of inline
+        /// elements, even if items are removed to make the result smaller than four elements.
+        /// </remarks>
+        private ArrayBuilder<T>? _builder;
+
+        private TemporaryArray(in TemporaryArray<T> array)
+        {
+            // Intentional copy used for creating an enumerator
+#pragma warning disable RS0042 // Do not copy value
+            this = array;
+#pragma warning restore RS0042 // Do not copy value
+        }
+
+        public static TemporaryArray<T> Empty => default;
+
+        public readonly int Count => _builder?.Count ?? _count;
+
+        public T this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            readonly get
+            {
+                if (_builder is not null)
+                    return _builder[index];
+
+                if ((uint)index >= _count)
+                    ThrowIndexOutOfRangeException();
+
+                return Unsafe.Add(ref Unsafe.AsRef(in _item0), index);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set
+            {
+                if (_builder is not null)
+                {
+                    _builder[index] = value;
+                    return;
+                }
+
+                if ((uint)index >= _count)
+                    ThrowIndexOutOfRangeException();
+
+                Unsafe.Add(ref _item0, index) = value;
+            }
+        }
+
+        public void Dispose()
+        {
+            // Make sure each inline item has a reference since they are accessed through address manipulation
+            GC.KeepAlive(_item0);
+            GC.KeepAlive(_item1);
+            GC.KeepAlive(_item2);
+            GC.KeepAlive(_item3);
+
+            Interlocked.Exchange(ref _builder, null)?.Free();
```

Comment:
can you either null out these four items, or explicitly state that we don't think it's necessary for yadda yadda.

Reply:
➡️ Commented

Reply:
thanks.